File: app/src/main/java/io/github/zwieback/familyfinance/business/operation/activity/TransferOperationEditActivity.java
package io.github.zwieback.familyfinance.business.operation.activity;

import android.app.Activity;
import android.content.Intent;
import android.view.View;
import android.widget.EditText;

import com.johnpetitto.validator.ValidatingTextInputLayout;
import com.mikepenz.iconics.view.IconicsImageView;

import org.droidparts.widget.ClearableEditText;

import java.util.Arrays;
import java.util.List;

import io.github.zwieback.familyfinance.R;
import io.github.zwieback.familyfinance.business.operation.adapter.TransferOperationProvider;
import io.github.zwieback.familyfinance.core.adapter.EntityProvider;
import io.github.zwieback.familyfinance.core.model.Account;
import io.github.zwieback.familyfinance.core.model.Article;
import io.github.zwieback.familyfinance.core.model.Operation;
import io.github.zwieback.familyfinance.core.model.type.OperationType;
import io.github.zwieback.familyfinance.databinding.ActivityEditTransferOperationBinding;
import io.reactivex.functions.Consumer;

import static io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.EXPENSE_ACCOUNT_CODE;
import static io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.INCOME_ACCOUNT_CODE;
import static io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.RESULT_ACCOUNT_ID;

public class TransferOperationEditActivity
        extends OperationEditActivity<ActivityEditTransferOperationBinding> {

    public static final String INPUT_TRANSFER_OPERATION_ID = "transferExpenseOperationId";
    public static final String INPUT_EXPENSE_ACCOUNT_ID = "expenseAccountId";
    public static final String OUTPUT_TRANSFER_OPERATION_ID = "resultTransferExpenseOperationId";

    /**
     * internal entity is alias for expense operation
     */
    private Operation incomeOperation;

    @Override
    protected int getTitleStringId() {
        return R.string.transfer_operation_activity_edit_title;
    }

    @Override
    protected int getBindingLayoutId() {
        return R.layout.activity_edit_transfer_operation;
    }

    @Override
    protected String getExtraInputId() {
        return INPUT_TRANSFER_OPERATION_ID;
    }

    @Override
    protected String getExtraOutputId() {
        return OUTPUT_TRANSFER_OPERATION_ID;
    }

    @Override
    protected EntityProvider<Operation> createProvider() {
        return new TransferOperationProvider(this);
    }

    @Override
    OperationType getOperationType() {
        return OperationType.TRANSFER_EXPENSE_OPERATION;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent resultIntent) {
        super.onActivityResult(requestCode, resultCode, resultIntent);
        switch (requestCode) {
            case EXPENSE_ACCOUNT_CODE:
                if (resultCode != Activity.RESULT_OK) {
                    break;
                }
                int expenseAccountId = extractOutputId(resultIntent, RESULT_ACCOUNT_ID);
                loadExpenseAccount(expenseAccountId);
                break;
            case INCOME_ACCOUNT_CODE:
                if (resultCode != Activity.RESULT_OK) {
                    break;
                }
                int incomeAccountId = extractOutputId(resultIntent, RESULT_ACCOUNT_ID);
                loadIncomeAccount(incomeAccountId);
                break;
        }
    }

    public void onExpenseAccountClick(View view) {
        startAccountActivity(EXPENSE_ACCOUNT_CODE);
    }

    public void onIncomeAccountClick(View view) {
        startAccountActivity(INCOME_ACCOUNT_CODE);
    }

    private Consumer<Account> onSuccessfulExpenseAccountFound() {
        return foundAccount -> {
            entity.setAccount(foundAccount);
            binding.expenseAccount.setText(foundAccount.getName());
            if (foundAccount.getOwner() != null) {
                loadOwner(foundAccount.getOwner().getId());
            }
        };
    }

    private Consumer<Account> onSuccessfulIncomeAccountFound() {
        return foundAccount -> {
            incomeOperation.setAccount(foundAccount);
            binding.incomeAccount.setText(foundAccount.getName());
        };
    }

    private Consumer<Article> onSuccessfulArticleFound() {
        return foundArticle -> entity.setArticle(foundArticle);
    }

    private void loadExpenseAccount(int accountId) {
        loadEntity(Account.class, accountId, onSuccessfulExpenseAccountFound());
    }

    private void loadIncomeAccount(int accountId) {
        loadEntity(Account.class, accountId, onSuccessfulIncomeAccountFound());
    }

    private void loadDefaultArticle() {
        int transferArticleId = databasePrefs.getTransferArticleId();
        loadEntity(Article.class, transferArticleId, onSuccessfulArticleFound());
    }

    private int extractExpenseAccountId() {
        return extractInputId(INPUT_EXPENSE_ACCOUNT_ID, databasePrefs.getAccountId());
    }

    @Override
    protected void createEntity() {
        super.createEntity();
        loadExpenseAccount(extractExpenseAccountId());
        loadDefaultArticle();
        createIncomeOperation();
    }

    private void createIncomeOperation() {
        incomeOperation = new Operation();
        incomeOperation.setType(OperationType.TRANSFER_INCOME_OPERATION);
        bindIncomeOperation(incomeOperation);
    }

    @Override
    protected void loadEntity(int expenseOperationId) {
        loadEntity(Operation.class, expenseOperationId, this::onSuccessfulExpenseOperationFound);
    }

    private void onSuccessfulExpenseOperationFound(Operation expenseOperation) {
        bind(expenseOperation);
        loadEntity(Operation.class, expenseOperation.getLinkedTransferOperation().getId(),
                this::bindIncomeOperation);
    }

    @Override
    protected void bind(Operation operation) {
        entity = operation;
        binding.setExpenseOperation(operation);
        provider.setupIcon(binding.icon.getIcon(), operation);
        super.bind(operation);
    }

    private void bindIncomeOperation(Operation operation) {
        incomeOperation = operation;
        binding.setIncomeOperation(operation);
        setupIncomeOperationBindings();
    }

    @Override
    protected void setupBindings() {
        binding.icon.setOnClickListener(this::onSelectIconClick);
        binding.owner.setOnClickListener(this::onOwnerClick);
        binding.expenseAccount.setOnClickListener(this::onExpenseAccountClick);
        binding.expenseAccount.setOnClearTextListener(() -> entity.setAccount(null));
        binding.date.setOnClickListener(this::onDateClick);
        binding.currency.setOnClickListener(this::onCurrencyClick);
        binding.exchangeRate.setOnClickListener(this::onExchangeRateClick);
        super.setupBindings();
    }

    private void setupIncomeOperationBindings() {
        binding.incomeAccount.setOnClickListener(this::onIncomeAccountClick);
        binding.incomeAccount.setOnClearTextListener(() -> incomeOperation.setAccount(null));
    }

    @Override
    protected void updateEntityProperties(Operation operation) {
        super.updateEntityProperties(operation);
        incomeOperation.setArticle(operation.getArticle());
        incomeOperation.setOwner(operation.getOwner());
        incomeOperation.setExchangeRate(operation.getExchangeRate());
        incomeOperation.setDate(operation.getDate());
        incomeOperation.setValue(operation.getValue());
        incomeOperation.setDescription(operation.getDescription());
    }

    @Override
    protected List<ValidatingTextInputLayout> getLayoutsForValidation() {
        return Arrays.asList(
                binding.ownerLayout,
                binding.expenseAccountLayout, binding.incomeAccountLayout,
                binding.valueLayout, binding.dateLayout,
                binding.currencyLayout, binding.exchangeRateLayout);
    }

    @Override
    protected Consumer<Operation> onSuccessfulSaving() {
        return expenseOperation -> {
            incomeOperation.setLinkedTransferOperation(expenseOperation);
            saveEntity(incomeOperation, onSuccessfulIncomeOperationSaving(expenseOperation));
        };
    }

    private Consumer<Operation> onSuccessfulIncomeOperationSaving(Operation expenseOperation) {
        return incomeOperation -> {
            expenseOperation.setLinkedTransferOperation(incomeOperation);
            saveEntity(expenseOperation, onSuccessfulExpenseOperationSaving());
        };
    }

    private Consumer<Operation> onSuccessfulExpenseOperationSaving() {
        return this::closeActivity;
    }

    @Override
    protected IconicsImageView getIconView() {
        return binding.icon;
    }

    @Override
    ClearableEditText getOwnerEdit() {
        return binding.owner;
    }

    @Override
    ClearableEditText getCurrencyEdit() {
        return binding.currency;
    }

    @Override
    ClearableEditText getExchangeRateEdit() {
        return binding.exchangeRate;
    }

    @Override
    EditText getDateEdit() {
        return binding.date;
    }

    @Override
    EditText getValueEdit() {
        return binding.value;
    }

    @Override
    EditText getDescriptionEdit() {
        return binding.description;
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/operation/activity/TransferOperationEditActivity.kt
package io.github.zwieback.familyfinance.business.operation.activity

import android.app.Activity
import android.content.Intent
import android.widget.EditText
import com.johnpetitto.validator.ValidatingTextInputLayout
import com.mikepenz.iconics.view.IconicsImageView
import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.Companion.EXPENSE_ACCOUNT_CODE
import io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.Companion.INCOME_ACCOUNT_CODE
import io.github.zwieback.familyfinance.business.dashboard.activity.DashboardActivity.Companion.RESULT_ACCOUNT_ID
import io.github.zwieback.familyfinance.business.operation.service.provider.TransferOperationProvider
import io.github.zwieback.familyfinance.core.adapter.EntityProvider
import io.github.zwieback.familyfinance.core.model.Account
import io.github.zwieback.familyfinance.core.model.Article
import io.github.zwieback.familyfinance.core.model.Operation
import io.github.zwieback.familyfinance.core.model.type.OperationType
import io.github.zwieback.familyfinance.databinding.ActivityEditTransferOperationBinding
import io.github.zwieback.familyfinance.extension.isEmptyId
import io.github.zwieback.familyfinance.widget.ClearableEditText
import io.reactivex.functions.Consumer
import org.threeten.bp.LocalDate
import org.threeten.bp.LocalDateTime
import java.math.BigDecimal

class TransferOperationEditActivity :
    OperationEditActivity<ActivityEditTransferOperationBinding>() {

    /**
     * internal [entity] is alias for expense operation
     */
    private lateinit var incomeOperation: Operation

    override val titleStringId: Int
        get() = R.string.transfer_operation_activity_edit_title

    override val bindingLayoutId: Int
        get() = R.layout.activity_edit_transfer_operation

    override val extraInputId: String
        get() = INPUT_TRANSFER_OPERATION_ID

    override val extraOutputId: String
        get() = OUTPUT_TRANSFER_OPERATION_ID

    override val operationType: OperationType
        get() = OperationType.TRANSFER_EXPENSE_OPERATION

    override val layoutsForValidation: List<ValidatingTextInputLayout>
        get() = listOf(
            binding.ownerLayout,
            binding.expenseAccountLayout,
            binding.incomeAccountLayout,
            binding.valueLayout,
            binding.dateLayout,
            binding.currencyLayout,
            binding.exchangeRateLayout
        )

    override val iconView: IconicsImageView
        get() = binding.icon

    override val ownerEdit: ClearableEditText
        get() = binding.owner

    override val currencyEdit: ClearableEditText
        get() = binding.currency

    override val exchangeRateEdit: ClearableEditText
        get() = binding.exchangeRate

    override val dateEdit: EditText
        get() = binding.date

    override val valueEdit: EditText
        get() = binding.value

    override val descriptionEdit: EditText
        get() = binding.description

    override val urlEdit: EditText
        get() = binding.url

    override fun createProvider(): EntityProvider<Operation> {
        return TransferOperationProvider(this)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, resultIntent: Intent?) {
        super.onActivityResult(requestCode, resultCode, resultIntent)
        if (resultCode != Activity.RESULT_OK) {
            return
        }
        when (requestCode) {
            EXPENSE_ACCOUNT_CODE -> resultIntent?.let {
                val expenseAccountId = extractOutputId(resultIntent, RESULT_ACCOUNT_ID)
                loadExpenseAccount(expenseAccountId)
            }
            INCOME_ACCOUNT_CODE -> resultIntent?.let {
                val incomeAccountId = extractOutputId(resultIntent, RESULT_ACCOUNT_ID)
                loadIncomeAccount(incomeAccountId)
            }
        }
    }

    private fun onExpenseAccountClick() {
        startAccountActivity(EXPENSE_ACCOUNT_CODE)
    }

    private fun onIncomeAccountClick() {
        startAccountActivity(INCOME_ACCOUNT_CODE)
    }

    private fun onSuccessfulExpenseAccountFound(): Consumer<Account> {
        return Consumer { foundAccount ->
            entity.setAccount(foundAccount)
            foundAccount.owner?.id?.let { ownerId -> loadOwner(ownerId) }
        }
    }

    private fun onSuccessfulIncomeAccountFound(): Consumer<Account> {
        return Consumer { foundAccount -> incomeOperation.setAccount(foundAccount) }
    }

    private fun onSuccessfulArticleFound(): Consumer<Article> {
        return Consumer { foundArticle -> entity.setArticle(foundArticle) }
    }

    private fun loadExpenseAccount(accountId: Int) {
        loadEntity(Account::class.java, accountId, onSuccessfulExpenseAccountFound())
    }

    private fun loadIncomeAccount(accountId: Int) {
        loadEntity(Account::class.java, accountId, onSuccessfulIncomeAccountFound())
    }

    private fun loadDefaultArticle() {
        val transferArticleId = databasePrefs.transferArticleId
        loadEntity(Article::class.java, transferArticleId, onSuccessfulArticleFound())
    }

    private fun extractExpenseAccountId(): Int {
        return extractInputId(INPUT_EXPENSE_ACCOUNT_ID, databasePrefs.accountId)
    }

    private fun extractIncomeAccountId(): Int {
        return extractInputId(INPUT_INCOME_ACCOUNT_ID)
    }

    private fun extractExpenseOwnerId(): Int {
        return extractInputId(INPUT_EXPENSE_OWNER_ID, databasePrefs.personId)
    }

    private fun extractExpenseCurrencyId(): Int {
        return extractInputId(INPUT_EXPENSE_CURRENCY_ID, databasePrefs.currencyId)
    }

    private fun extractExpenseExchangeRateId(): Int {
        return extractInputId(INPUT_EXPENSE_EXCHANGE_RATE_ID)
    }

    private fun extractExpenseDate(): LocalDate {
        return extractInputDate(INPUT_EXPENSE_DATE)
    }

    private fun extractExpenseValue(): BigDecimal? {
        return extractInputBigDecimal(INPUT_EXPENSE_VALUE)
    }

    private fun extractExpenseDescription(): String? {
        return extractInputString(INPUT_EXPENSE_DESCRIPTION)
    }

    private fun extractExpenseUrl(): String? {
        return extractInputString(INPUT_EXPENSE_URL)
    }

    override fun createEntity() {
        super.createEntity()
        createIncomeOperation()
        loadDefaultArticle()
        loadExpenseAccount(extractExpenseAccountId())
        loadIncomeAccount(extractIncomeAccountId())
        loadOwner(extractExpenseOwnerId())
        val exchangeRateId = extractExpenseExchangeRateId()
        if (exchangeRateId.isEmptyId()) {
            loadCurrency(extractExpenseCurrencyId())
        } else {
            loadExchangeRate(exchangeRateId)
        }
    }

    override fun createOperation(): Operation {
        return super.createOperation()
            .setDate(extractExpenseDate())
            .setValue(extractExpenseValue())
            .setDescription(extractExpenseDescription())
            .setUrl(extractExpenseUrl())
    }

    private fun createIncomeOperation() {
        incomeOperation = Operation()
            .setCreateDate(LocalDateTime.now())
            .setType(OperationType.TRANSFER_INCOME_OPERATION)
        bindIncomeOperation(incomeOperation)
    }

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun loadEntity(expenseOperationId: Int) {
        loadEntity(
            Operation::class.java,
            expenseOperationId
        ) { onSuccessfulExpenseOperationFound(it) }
    }

    private fun onSuccessfulExpenseOperationFound(expenseOperation: Operation) {
        bind(expenseOperation)
        loadEntity(
            Operation::class.java,
            expenseOperation.linkedTransferOperation?.id ?: error("No linked income operation")
        ) { bindIncomeOperation(it) }
    }

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun bind(operation: Operation) {
        entity = operation
        binding.expenseOperation = operation
        provider.setupIcon(binding.icon.icon, operation)
        super.bind(operation)
    }

    private fun bindIncomeOperation(operation: Operation) {
        incomeOperation = operation
        binding.incomeOperation = operation
        setupIncomeOperationBindings()
    }

    override fun setupBindings() {
        binding.icon.setOnClickListener { onSelectIconClick() }
        binding.owner.setOnClickListener { onOwnerClick() }
        binding.expenseAccount.setOnClickListener { onExpenseAccountClick() }
        binding.expenseAccount.setOnClearTextListener { entity.setAccount(null) }
        binding.date.setOnClickListener { onDateClick() }
        binding.currency.setOnClickListener { onCurrencyClick() }
        binding.exchangeRate.setOnClickListener { onExchangeRateClick() }
        super.setupBindings()
    }

    private fun setupIncomeOperationBindings() {
        binding.incomeAccount.setOnClickListener { onIncomeAccountClick() }
        binding.incomeAccount.setOnClearTextListener { incomeOperation.setAccount(null) }
    }

    override fun updateEntityProperties(operation: Operation) {
        super.updateEntityProperties(operation)
        incomeOperation.setLastChangeDate(LocalDateTime.now())
        incomeOperation.setArticle(operation.article)
        incomeOperation.setOwner(operation.owner)
        incomeOperation.setExchangeRate(operation.exchangeRate)
        incomeOperation.setDate(operation.date)
        incomeOperation.setValue(operation.value)
        incomeOperation.setDescription(operation.description)
        incomeOperation.setUrl(operation.url)
    }

    override fun onSuccessfulSaving(): Consumer<Operation> {
        return Consumer { expenseOperation ->
            incomeOperation.setLinkedTransferOperation(expenseOperation)
            saveEntity(incomeOperation, onSuccessfulIncomeOperationSaving(expenseOperation))
        }
    }

    private fun onSuccessfulIncomeOperationSaving(expenseOperation: Operation): Consumer<Operation> {
        return Consumer { incomeOperation ->
            expenseOperation.setLinkedTransferOperation(incomeOperation)
            saveEntity(expenseOperation, onSuccessfulExpenseOperationSaving())
        }
    }

    private fun onSuccessfulExpenseOperationSaving(): Consumer<Operation> {
        return Consumer { closeActivity(it) }
    }

    companion object {
        const val INPUT_TRANSFER_OPERATION_ID = "transferExpenseOperationId"
        const val INPUT_EXPENSE_ACCOUNT_ID = "expenseAccountId"
        const val INPUT_INCOME_ACCOUNT_ID = "incomeAccountId"
        const val INPUT_EXPENSE_OWNER_ID = "expenseOwnerId"
        const val INPUT_EXPENSE_CURRENCY_ID = "expenseCurrencyId"
        const val INPUT_EXPENSE_EXCHANGE_RATE_ID = "expenseExchangeRateId"
        const val INPUT_EXPENSE_VALUE = "expenseValue"
        const val INPUT_EXPENSE_DATE = "expenseDate"
        const val INPUT_EXPENSE_DESCRIPTION = "expenseDescription"
        const val INPUT_EXPENSE_URL = "expenseUrl"
        const val OUTPUT_TRANSFER_OPERATION_ID = "resultTransferExpenseOperationId"
    }
}
